### ISSUE TITLE
Proposal: Replace _.kebabCase() with a different function

### ISSUE THREAD

**Issue by chmac** [gatsbyjs/gatsby#4637]
I believe a page's `path` is converted to its file name via `_.kebabCase()` [in](https://github.com/gatsbyjs/gatsby/blob/47a35aad300b9789ed328146fb71449652f4d101/packages/gatsby/cache-dir/static-entry.js#L28) [a](https://github.com/gatsbyjs/gatsby/blob/47a35aad300b9789ed328146fb71449652f4d101/packages/gatsby/src/redux/actions.js#L232) [few](https://github.com/gatsbyjs/gatsby/blob/47a35aad300b9789ed328146fb71449652f4d101/packages/gatsby/src/utils/js-chunk-names.js#L6) [places](https://github.com/gatsbyjs/gatsby/blob/47a35aad300b9789ed328146fb71449652f4d101/packages/gatsby-plugin-netlify/src/build-headers-program.js#L31). This results in transformations like so:

* `/foo` -> `foo`
* `/_foo` -> `foo`
* `/foo/bar` -> `foo-bar`
* `/foo_bar` -> `foo-bar`
* `/_foo/_bar` -> `foo-bar`

In my quick and dirty testing creating a page `/foo` and another page `/_foo` results in the wrong data being supplied to `/_foo`.

## Proposal

My suggestion would be to replace `_.kebabCase()` with a function which produces a unique output for every possible URL.

## Notes

NOTE: I think this will have far reaching consequences because I believe there are multiple places in the codebase which assume `_.kebabCase(path)` will return the correct filename.

Personally I only noticed this by digging through the code, I haven't actually faced this issue, so it's probably a low priority change.

**Comment by KyleAMathews** [comment]
Oh hmmm — you're right. Any suggestions on a better option?

**Comment by KyleAMathews** [comment]
Yeah, would love it if you could do some research!

**Comment by chmac** [comment]
Turns out this is more complicated than I initially thought. First, the path part of a URL is case sensitive. The majority of macOS machines (presumably where most Gatsby development is done) have a case insensitive filesystem by default. Oh joy.

Second, URL paths permit alphanumerics `[0-9a-zA-Z]` and some special characters characters `$-_.+!*'(),` without encoding. Plus the `%` to allow encoding.

Having put some thought into this, I see two possible approaches:
* Use a very complicated encoding scheme to generate filenames
  - First thought `encodeURIComponent()`, but it doesn't work.
  - Gets very complicated very fast. :-(
* Only allow something like `[0-9a-z\-]` (alphanumerics and the `-`) in paths
  - Would be a breaking change.

Right now you can have a path like `/fOO/_bar/baz(*)_/` and Gatsby will work just fine. This will be considered the same path as `/foo/bar/baz/`, but so long as you don't have any collisions like that, all should be well. Restricting the allowed characters in paths would be a breaking change.

@KyleAMathews I guess this is a topic on which you probably need to make a judgement call. There isn't an obviously better option than what we have right now. What do you think? Happy to provide more context, share more of the research / reasoning / etc if that's helpful, just let me know.

**Comment by KyleAMathews** [comment]
Perhaps we should just validate against paths that overlap? And use another casing scheme that doesn't make _foo and foo conflict.

**Comment by chmac** [comment]
Hmm, that's an interesting idea. 

I was coming at this from #4626. That lead me to  think about a deterministic way of converting a `path` to an ID and back. But if I put that goal aside for a second (it's probably impossible anyway), then collision detection would definitely be possible. We could use something like [github-slugger](https://www.npmjs.com/package/github-slugger) to get IDs like `foo`, `foo-1`, etc.

## Imlementation

Any initial thoughts on how we'd implement this? The `json-loader` branch in #4555 adds a `jsonName` value to all pages in redux. Maybe we do something similar and add the page ID to redux? Then we could handle collisions in the reducer?

Question: Does it matter if the IDs of each page change per build? I guess that if we detect collisions, the order of calls to `createPage()` will be significant. I further guess that the order may change between builds.

**Comment by KyleAMathews** [comment]
IDs do matter for long-term caching — so we wouldn't want the order of calls to be significant.

How about instead we keep using kebabCase but we hash the original file name and use the first 3 characters from the resulting hash and append that to the jsonName? That'd be enough to differentiate between any name collisions kebabCase causes.

Thanks for noticing this! The problem you identified is bound to hit someone sooner or later (if it isn't already causing trouble) and would be a very hard bug to identify.

**Comment by chmac** [comment]
OK, have thought about this a bunch. Removing the need to go from ID -> `path` makes life simpler. Unfortunately `encodeURIComponent()` doesn't really work because it leaves many characters unencoded. But I think it's the right general idea.

I suggest we use something like this:

```es6
// @flow

import _ from "lodash"

const allowed = "abcdefghijklmnopqrstuvwxyz0123456789"

const upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

const isUpper = (char: string): boolean => _.includes(upper, char)

const isAllowed = (char: string): boolean => _.includes(allowed, char)

const encodeUpper = (char: string): string => "u" + char.toLowerCase()

const encodeSymbol = (char: string): string =>
  char === "/" ? "-" : "c" + char.charCodeAt()

const encodeChar = (char: string): string =>
  isUpper(char) ? encodeUpper(char) : encodeSymbol(char)

const maybeEncode = (char: string): string =>
  isAllowed(char) ? char : encodeChar(char)

const removeLeadingSlash = (path: string): string => _.trimStart(path, "/")

const mapMaybeEncode = (path: string): string =>
  _.map(path, maybeEncode).join("")

const convertPathToId = _.flow([removeLeadingSlash, mapMaybeEncode])
```

This should produce outputs like:

```
   convertPathToId('/foo')
=> 'foo'
   convertPathToId('/_foo')
=> 'c95foo'
   convertPathToId('/_foo/BAR')
=> 'c95foo-ubuaur'
   convertPathToId('/_foo/BAR/__admin/$$')
=> 'c95foo-ubuaur-c95c95admin-c36c36'
```

It's somewhat involved. It'd be ideal if we could find a simpler solution. I would have preferred to use `lodash/fp`, but I guess that makes the code less readable. :-)

What do you think? All constructive feedback welcome.

I'm happy to start to work up a PR if this approach makes sense. I guess it makes sense to target `v2` as the changes could be far reaching. There seem to be a lot of places that use `_.kebabCase()` directly.

**Comment by KyleAMathews** [comment]
A simpler algorithm that I think would work is:

1) Do normal kebabCase conversion
2) if there's any characters that'd be changed or removed, add an encoded snippet at the end to differentiate it from other potential collisions.

This way we wouldn't need to add extra encoded stuff. We just detect if it could be colliding and then add the shortened hash at the end to prevent this.

**Comment by KyleAMathews** [comment]
How is this a breaking change?

Yeah exactly, we test if the string has a character that'll be changed and if so, append the hash.

**Comment by chmac** [comment]
> How is this a breaking change?

Maybe "breaking change" is the wrong term. It's going to change some IDs, so a build with the previous version and this version will result in different output. It'll cause some cache burn. If this change lands in v2, then I guess it's a moot point.

> Yeah exactly, we test if the string has a character that'll be changed and if so, append the hash.

I made a mistake in my last comment. The only paths that won't be a potential collision would contain numbers, lower case letters, and forward slashes. Anything else, would result in a hash being appended. So `/foo-bar` would result in a hash being added. So any path with more than one word will get a new ID. `/about/` doesn't change, but `/about-us/` does.

What's your thinking behind preferring the kebabCase + hash approach? Happy to implement it, and I defer to your judgement, you're the Gatsby expert! Hoping to better understand your thinking.

**Comment by KyleAMathews** [comment]
>What's your thinking behind preferring the kebabCase + hash approach?

It's simple to implement and understand. Is there a better way or reason not to do it this way?

**Comment by chmac** [comment]
> It's simple to implement and understand.

I hear ya. The simplicity of the output is clear. Under "normal" situations, the IDs will definitely be more readable than the weird encoding stuff I came up with.

> Is there a better way or reason not to do it this way?

I see a few arguments against it.

## Collisions

The first 3 characters of a hash create the potential for collisions. It's unlikely, assuming a hex representation of the hash, 16^3, so a 1 in 4096 chance of collision. Low, but not impossible.

## Readability

It's not possible to tell whether `foo-bar-XXX` or `foo-bar-YYY` is the ID for `/foo/bar/` or `/FooBar/`, while it would be possible with the encoding method. The encoded IDs are definitely unwieldy, but they can theoretically be parsed by a human, while a cryptographic hash cannot.

## Hashing in the browser

How widely is hashing available in the browser? I saw something about crypto functions being added in 2017, but no idea how widely available they are. My assumption is that we need a polyfill or a library. The encoding method's only dependency is lodash.

Personally, the collision argument is the biggest one. I feel some internal resistance at the idea of solving the problem with a method that makes the error less likely but not impossible! :-)

**Comment by KyleAMathews** [comment]
> Collisions

We could pop that up to 4. But collisions are already very unlikely given people would have to name their pages nearly identically but in a way that the kebab-case causes the collisions. Which afaik, hasn't happened yet in the wild.

> Hashing in the browser

This is just for generating path names during the site build so none of this is executed in the browser.

**Comment by chmac** [comment]
True enough. Hash collisions would only be on clashing paths. Very unlikely, you're right.

I'm thinking we might need to generate IDs on the client. We don't today, you're right there. If we start sharding `data.json` we probably will. But the hash polyfill argument is a pretty weak one anyway.

Think you've sold me, the kebab case + hash makes sense. It's definitely more natural to read.

Any reason not to add the hash to all IDs?

I'm guessing (totally unsubstantiated) that 80% of `path`s will contain a dash and so will get a hash anyway. Maybe simpler to have the same format for all IDs?

**Comment by KyleAMathews** [comment]
>I'm guessing (totally unsubstantiated) that 80% of paths will contain a dash and so will get a hash anyway. Maybe simpler to have the same format for all IDs?

That's a good point — might as well just add a hash to all of them. That'll make the logic even simpler. The simpler the implementation the better.

**Comment by chmac** [comment]
Quick update to say we're swamped work wise the next 3 weeks, but I've put this on my list to get stuck into after that.

**Comment by KyleAMathews** [comment]
Thanks for the update! Looking forward to this landing.

**Comment by chmac** [comment]
Fixed by #5532
